lib/CGameState.cpp

```diff
@@ -1608,6 +1610,64 @@ void CGameState::giveCampaignBonusToHero(CGHeroInstance * hero)
 	}
 }
 
+void CGameState::removeSightnObj(const CGObjectInstance * obj)
+{
+	addSightObj(obj, false);
+}
+
+void CGameState::addSightObj(const CGObjectInstance * obj, bool add)
+{
+	if(!vstd::contains(players, obj->tempOwner))
+		return;
+
+	auto p = getPlayer(obj->tempOwner);
+	if(p->status != EPlayerStatus::INGAME)
+		return;
+
+	addSightObj(p->team, obj, add);
+}
+
+void CGameState::addSightObj(TeamID team, const CGObjectInstance * obj, bool add)
+{
+	logGlobal->debug("%s object on sight map: %d", add ? "Addding" : "Removing", obj->id);
+	auto ts = getTeam(team);
+	std::unordered_set<int3, ShashInt3> tiles;
+	getTilesInRange(tiles, obj->getSightCenter(), obj->getSightRadius());
+	for(int3 t : tiles)
+	{
+		auto old = ts->fogOfWarMap[t.x][t.y][t.z];
+		if(add)
+		{
+			if(ts->fogOfWarMap[t.x][t.y][t.z] == 0)
+				ts->fogOfWarMap[t.x][t.y][t.z] = 2;
+			else
+				ts->fogOfWarMap[t.x][t.y][t.z]++;
```

Comment:
If it's happen that would mean FoW going to break completely.
So I can just add assert so it's indicate if something going really wrong.